File: tests/test_plugin_trust_classify.py
"""Regression for the trust classifier (dev_server.py:_classify_plugin) covering the
P2-4 + audit fixes: sandbox:true routes to the sandbox lane regardless of trust tier,
sandboxed-approval overrides dev-trust, cap-subset + hash-pinning void a stale approval.
Twin: tests/test_plugin_trust_classify.php. Uses the bundled fixtures."""
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
import dev_server as ds

SB = ds.MODULES_DIR / "tools/screenshot-sandboxed"   # sandbox:true
REG = ds.MODULES_DIR / "tools/screenshot"            # regular in-page
H = ds._plugin_hash(ds._plugin_file_hashes(SB))
DECLARED = sorted(ds._plugin_declared_caps(SB))
SBP, REGP = "tools/screenshot-sandboxed", "tools/screenshot"
fails = 0


def check(name, cond):
    global fails
    print(("  ok   " if cond else "  FAIL ") + name)
    if not cond:
        fails += 1


sandbox_ap = [{"path": SBP, "sha256": H, "mode": "sandboxed", "caps": DECLARED}]

# `sandbox: true` decides the LANE — always the iframe, whatever the trust tier.
ds._DEV_TRUST = True
check("dev + sandbox:true plugin -> sandboxed (author lane)", ds._classify_plugin(SBP, SB, [], None)["tier"] == "sandboxed")
check("dev + regular plugin -> dev (in-page)", ds._classify_plugin(REGP, REG, [], None)["tier"] == "dev")

ds._DEV_TRUST = False
check("prod: sandboxed approval -> sandboxed", ds._classify_plugin(SBP, SB, sandbox_ap, None)["tier"] == "sandboxed")
check("prod: no approval -> untrusted", ds._classify_plugin(SBP, SB, [], None)["tier"] == "untrusted")
check("prod: regular no approval -> untrusted", ds._classify_plugin(REGP, REG, [], None)["tier"] == "untrusted")
check("prod: hash drift voids approval",
      ds._classify_plugin(SBP, SB, [{"path": SBP, "sha256": "0" * 64, "mode": "sandboxed", "caps": DECLARED}], None)["tier"] == "untrusted")
check("prod: under-granted caps void approval",
      ds._classify_plugin(SBP, SB, [{"path": SBP, "sha256": H, "mode": "sandboxed", "caps": ["ui.toast"]}], None)["tier"] == "untrusted")

if fails:
    print(f"\n{fails} CLASSIFY CHECKS FAILED")
    sys.exit(1)
print("\nALL TRUST-CLASSIFY CHECKS PASSED (python)")

File: tests/test_plugin_trust_classify.php
<?php
/* PHP twin of tests/test_plugin_trust_classify.py — asserts admin_classify_plugin
   agrees with dev_server.py:_classify_plugin on the key semantics: sandbox:true
   routes to the sandbox lane, dev-trust is loopback-gated, sandboxed-approval wins,
   hash-pinning + cap-subset void a stale approval. Uses the bundled fixtures. */
declare(strict_types=1);
require_once __DIR__ . '/../api/_admin_lib.php';

$_SERVER['REMOTE_ADDR'] = '127.0.0.1';  // loopback → dev-trust active (a .git checkout)
$SB  = modules_dir() . '/tools/screenshot-sandboxed';  // sandbox:true
$REG = modules_dir() . '/tools/screenshot';            // regular in-page
$hSB = admin_plugin_hash(admin_plugin_file_hashes($SB));
$declared = admin_plugin_declared_caps($SB);
$fails = 0;
function check($name, $cond) { global $fails; echo ($cond ? "  ok   " : "  FAIL ") . "$name\n"; if (!$cond) $fails++; }

check('dev + sandbox:true -> sandboxed (author lane)',
    admin_classify_plugin('tools/screenshot-sandboxed', $SB, [], null)['tier'] === 'sandboxed');
check('dev + regular -> dev (in-page)',
    admin_classify_plugin('tools/screenshot', $REG, [], null)['tier'] === 'dev');

$_SERVER['REMOTE_ADDR'] = '192.168.1.9';  // LAN client → NO dev-trust
check('LAN + regular -> untrusted (loopback gate)',
    admin_classify_plugin('tools/screenshot', $REG, [], null)['tier'] === 'untrusted');
check('LAN + sandboxed approval (hash ok) -> sandboxed (wins)',
    admin_classify_plugin('tools/screenshot-sandboxed', $SB,
        [['path' => 'tools/screenshot-sandboxed', 'sha256' => $hSB, 'mode' => 'sandboxed', 'caps' => $declared]], null)['tier'] === 'sandboxed');
check('LAN + hash drift -> untrusted',
    admin_classify_plugin('tools/screenshot-sandboxed', $SB,
        [['path' => 'tools/screenshot-sandboxed', 'sha256' => str_repeat('0', 64), 'mode' => 'sandboxed', 'caps' => $declared]], null)['tier'] === 'untrusted');
check('LAN + under-granted caps -> untrusted',
    admin_classify_plugin('tools/screenshot-sandboxed', $SB,
        [['path' => 'tools/screenshot-sandboxed', 'sha256' => $hSB, 'mode' => 'sandboxed', 'caps' => ['ui.toast']]], null)['tier'] === 'untrusted');

if ($fails) { echo "\n$fails CLASSIFY CHECKS FAILED (php)\n"; exit(1); }
echo "\nALL TRUST-CLASSIFY CHECKS PASSED (php)\n";
